### PR TITLE
fix schema definition for payload template to avoid the oneOf options intersecting

### LINKED ISSUE #104: Validation error for Parameters
**Describe the bug**
New error based on changes made in [101](https://github.com/ChristopheBougere/asl-validator/issues/101) / [102](https://github.com/ChristopheBougere/asl-validator/issues/102)
Validation error
```
 {
    'Error code': 'SCHEMA_VALIDATION_FAILED',
    Message: '/States/Publish to Slack/Parameters/slackMessage/blocks is invalid. must match exactly one schema in oneOf',
    schemaError: {
      instancePath: '/States/Publish to Slack/Parameters/slackMessage/blocks',
      schemaPath: '#/patternProperties/%5E.%2B((%5B%5E.%5D%5B%5E%24%5D)%7C(%5B%5E.%5D%5B%24%5D))%24/oneOf'
    }
  }
```

**To Reproduce**
Please provide a complete state machine definition:
```json
{
  "StartAt": "Publish to Slack",
  "States": {
      "Publish to Slack": {
      "Type": "Task",
      "Resource": "arn:aws:lambda:::function:publish-to-slack-lambda",
      "Parameters": {
        "slackMessage": {
          "channel.$": "$.slackErrorChannel",
          "text": "$.slackMessage.title",
          "blocks": [
            {
              "type": "section",
              "text": {
                "type": "mrkdwn",
                "text.$": "$.slackMessage.errorMessage"
              }
            },
            {
              "type": "section",
              "text": {
                "type": "mrkdwn",
                "text.$": "$.slackMessage.orderId"
              }
            },
            {
              "type": "section",
              "text": {
                "type": "mrkdwn",
                "text.$": "$.slackMessage.orderItemId"
              }
            },
            {
              "type": "section",
              "text": {
                "type": "mrkdwn",
                "text.$": "$.slackMessage.stack"
              }
            },
            {
              "type": "section",
              "text": {
                "type": "mrkdwn",
                "text.$": "$.slackMessage.link"
              }
            }
          ]
        }
      },
      "Next": "Failure"
    },
    "Failure": {
      "Type": "Fail"
    },
    "Success": {
      "Type": "Succeed"
    }
  }
}
```

**Expected behavior**
No validation error

**Version:**
3.0.2
**Additional context**

### PR DESCRIPTION
**Pull request by massfords**
AJV is failing the input schema since it is matching on two of the available types in its `oneOf` definition. 

closes #104 

fix(schemas): recursive def fixed where oneOf options intersected on array types
fix(check): catch jsonpath error in duplicate-payload-template-fields.ts
refactor(check): decode URI encoded values for cleaner error reporting

### CODE REVIEW

File: src/__tests__/definitions/valid-parameters-issue104.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+          "text.$": "$.slackMessage.title",
+          "blocks": [
+            {
+              "type": "section",
+              "text": {
+                "type": "mrkdwn",
+                "text.$": "$.slackMessage.errorMessage"
+              }
+            }
+          ]
```
Comment (massfords) [inline src/__tests__/definitions/valid-parameters-issue104.json:19]: this is sufficient input to trigger the error

the value for "blocks" was matching as an array type for two of the definitions in a `oneOf`

File: src/__tests__/error-reporting.test.ts [hunk 2]
```diff
@@ -43,8 +43,7 @@ describe("tests with definitions containing errors to see what's reported", () =
                     Message: "/States/Hello, World/Parameters/lorem.$ is invalid. must match format \"asl_payload_template\"",
                     schemaError: {
                         instancePath: "/States/Hello, World/Parameters/lorem.$",
-                        // the regular expression in the path is escaped by AJV when it reports the location of the error
-                        schemaPath: "#/patternProperties/%5E.%2B%5C.%5C%24%24/format"
+                        schemaPath: "#/patternProperties/^.+\\.\\$$/format"
```
Comment (massfords) [inline src/__tests__/error-reporting.test.ts:46]: unrelated change to decode the schemaPath in case it has encoded values.

File: src/checks/duplicate-payload-template-fields.ts [hunk 3]
```diff
@@ -9,33 +9,40 @@ export const mustNotHaveDuplicateFieldNamesAfterEvaluation: AslChecker = (defini
...
+            // also, don't miss the array offset [0] before selecting the grandparent.
+            // There may be multiple fields within a node that could be in conflict.
+            // We're checking all the keys in the parent node, so we only need a single
+            // result for that parent.
+            path: `$..[Parameters,ResultSelector]..*[?(@property.match(/^.+\\.\\$$/))][0]^^`
+        })
+    } catch (err: unknown) {
+        // need to revisit this since the JSONPath lib throws an error when it encounters an array
+        // Error: jsonPath: _$_property.match is not a function: _$_property.match(/^.+\.\$$/)
+    }
```
Comment (massfords) [inline src/checks/duplicate-payload-template-fields.ts:31]: I'll write a new issue for this and look at a better fix later.

After fixing the schema issue in this PR, the test failed with a runtime error from JSONPath Plus library as noted in the comment.

File: src/schemas/paths.json [hunk 4]
```diff
@@ -37,7 +37,7 @@
               }
             },
             {
-              "$ref": "#/definitions/_asl_payload_template"
+              "$ref": "#/definitions/_payload_template_object"
```
Comment (massfords) [inline src/schemas/paths.json:40]: changed to target the object instead of the abstraction which allows for an array as well.

File: src/checks/duplicate-payload-template-fields.ts (hunk 3 above)
Comment (ChristopheBougere) [reply src/checks/duplicate-payload-template-fields.ts:31]: Sounds good

**Review by ChristopheBougere — APPROVED** [review]
👍
FYI, the commit ling is failing because your commit messages don't have a leading blank line. 
Also, I think you're supposed to have a single semantic message per commit, otherwise they'll be squashed in a single one.
I'll squash and rewrite in the merge.
